util/llbutil/pllb/state.go

```diff
@@ -122,6 +142,67 @@ func (s State) GetDir(ctx context.Context) (string, error) {
 	return s.st.GetDir(ctx)
 }
 
+func getSharedKeyHintFromInclude(name string, incl []string) string {
+	h := sha1.New()
+	b := make([]byte, 8)
+
+	addToHash := func(path string) {
+		h.Write([]byte(path))
+		inode := getInodeBestEffort(path)
+		binary.LittleEndian.PutUint64(b, inode)
+		h.Write(b)
+	}
+
+	addToHash(name)
+	for _, path := range incl {
+		addToHash(path)
+	}
+	return hex.EncodeToString(h.Sum(nil))
+}
```

Comment:
I wouldn't put this kind of logic in this specific file. I think there's value for this file to remain a thin layer on top of `llb.State`, and keep things simple at this level.

Could this functionality be offered as a layer of top of the `llb.State` API and not within it?

Reply:
We could *easily* change `WithInclude(incl []string)` to `WithInclude(incl []string, key string)`.

Alternatively, If we want to break it up completely into `WithInclude()` as well as a `WithSharedKeyHint()`, then that change is a much larger refactor as we would have to delay calling `llb.Local()` until all of these chained modifier methods are called. Then perhaps have a final `Construct()` method.

That would change such a call into:

    state.Local(name).WithInclude(incl).WithSharedKeyHint(key).Construct()

But at that point I would question the use of this file completely, since it wouldn't seem that different than calling

    llb.Local(s.localName, opts...)

directly.

Thoughts?

Reply:
> We could easily change WithInclude(incl []string) to WithInclude(incl []string, key string).

I think this makes sense - better to be more explicit.

I'm also proposing to not put this functionality in this package and instead to have it in say `llbutil/localutil`. This

`func (s State) WithInclude(incl []string) State`

maybe becomes

`func StateWithInclude(in pllb.State, incl []string, key string) pllb.State`

in that package.

Reply:
Gotcha, If we do that, we'll need to add a way to determine if the `pllb.State` is local (and what it's localname is), for us to reconstruct it with the Include.

What's tricky is we construct the state under https://github.com/earthly/earthly/blob/local-include-pattern/buildcontext/local.go#L65-L70

Then it ultimately gets passed to a newInterpreter under https://github.com/earthly/earthly/blob/local-include-pattern/earthfile2llb/earthfile2llb.go#L124

But at that point we've lost access to if it's local or it's options; however we won't know what the COPY args until we reach the interpreter's handleCopy https://github.com/earthly/earthly/blob/local-include-pattern/earthfile2llb/interpreter.go#L614

This sort of goes back to the idea of having a factory type call where we chain on configuration methods onto it, and delay the construction until we eventually reach the converter https://github.com/earthly/earthly/blob/local-include-pattern/earthfile2llb/converter.go#L419 (or pre-construct it one level higher in the interpreter).

So ultimately do we build a factory, or just expose two `GetLocalName()` and `GetLocalOpts()` methods in pllb? And at that point are these two methods better or worse than having `WithInclude` in this package?

Reply:
I think it might be better as a factory then. I know it ends up being more code, but the advantage is that the behavior is more segregated, which helps to create lower coupling.

Plus, the `llb.State` API is then pretty much the same as Buildkit's - so there's no unexpected magic going on in the middle layers, which makes the code easier to understand.

Reply:
Hah, I was afraid of that :)

It'll allow us to be more flexible in the future. I'll see what can be done to switch it over to a factory.

Reply:
I created a factory with a specific `LocalFactory` to help us in this case, as well as a generic `PreconstructedFactory` which allows us to pass in a preconstructed pllb.State. I thought that was a good compramise on not implementing all possible factory types at this time, and if any other requirements come up later, we can add to the existing factory.